vermouth/tests/test_name_moltype.py

```diff
@@ -0,0 +1,151 @@
+
+@st.composite
+def molecules_and_moltypes(draw, max_moltypes=4, min_size=0, max_size=None,
+                           molecule_kwargs={}):
+    """
+    Generates a list of molecules and the list of the expected moltypes.
+
+    Parameters
+    ----------
+    draw:
+        Internal for hypothesis.
+    max_moltypes: int
+        Maximum number of different moltypes.
+    min_size: int
+        Minimum number of molecules.
+    max_size: int, optional
+        Maximum number of molecules. If `None`, behaves the same way as
+        :func:`hypothesis.strategies.lists`.
+    molecule_kwargs: dict
+        Arguments to pass to :func:`random_molecule`.
+
+    Returns
+    -------
+    hypothesis.searchstrategy.lazy.LazyStrategy
+        A list of molecules and a list of moltype names corresponding to these
+        molecules as they are expected to be assigned by
+        :class:`vermouth.processors.name_moltype.NameMolType`.
+    """
+    if max_size is not None and max_size < min_size:
+        raise ValueError('max_size ({}) must be greater than min_size ({}), or None.'
+                         .fornat(max_size, min_size))
+
+    # We first generate the list of expected moltypes, we can then use it as
+    # keys to draw the molecules with shared strategies. The moltypes must
+    # match what NameMolType would generate, so they must have the form
+    # molecule_0, molecule_1, molecule_x, where the index is increasing based
+    # on the first occurence of the molecule type.
+    # pre_moltypes will be the moltype names before we rename them based on
+    # their order.
+    n_moltypes = draw(st.integers(min_value=1, max_value=max_moltypes))
+    unique_pre_moltypes = range(n_moltypes)
+    pre_moltypes_sampled = draw(st.lists(st.sampled_from(unique_pre_moltypes),
+                                         min_size=min_size, max_size=max_size))
+    if not pre_moltypes_sampled:
+        return [], []
+
+    moltype_index = 0
+    pre_to_moltype = {pre_moltypes_sampled[0]: 'molecule_{}'.format(moltype_index)}
+    moltypes = []
+    for pre_moltype in pre_moltypes_sampled:
+        if pre_moltype not in pre_to_moltype:
+            moltype_index += 1
+            pre_to_moltype[pre_moltype] = 'molecule_{}'.format(moltype_index)
+        moltypes.append(pre_to_moltype[pre_moltype])
+
+    sharing = defaultdict(lambda : draw(random_molecule(**molecule_kwargs)))
+    molecules = [sharing[moltype] for moltype in moltypes]
+
+    # So far we made sure that molecules with a different moltype are drawn
+    # separately. Though, two different draws may result in equal molecules,
+    # especially as molecules are reduced whn shrinking the draws.
+    representatives = [
+        next(group)[0]
+        for _, group in itertools.groupby(
+                zip(molecules, moltypes), key=lambda x: x[1])
+    ]
+    assume(all(not mol1.share_moltype_with(mol2)
+               for mol1, mol2 in itertools.combinations(representatives, 2)))
```

Comment:
How often does this reject a sample? Hypothesis statistics should tell you.

Reply:
In the last travis run, the rejection rate is about 20 to 30%. It would be lower for larger molecules, but generating these larger molecules would be much longer.

Reply:
Acceptable. But it might be worth looking at a separate strategy that draws lists of unique elements (using provided strategy) that uses equality testing.

```python3
def unique_list(draw, strategy, min_size=0, max_size=None):
    output = []
    num_items = draw(st.integer(min=min_size, max=max_size)
    for _ in range(num_items):
        new_item = draw(strategy)
        assume(not any(item == new_item for item in output))
        output.append(item)
    return output

```

Reply:
It would be nice to have at some point indeed, but in this specific case, it would not reduce the rejection rate.

Reply:
Replace assume for:
```python
while True:
    new_item = draw(strategy)
    if not any(item == new_item for item in output):
        break
```

Reply:
I expect this to go rather badly with the shrinking. It would help me if you could tell me what you do not like with the `assume`: I do not understand what is the problem you are trying to solve.

Reply:
It's from the category of overengineering. 20-30% rejection is acceptable.

My problem with assume is that it'll reject the entire draw, instead of making a single new element.

Reply:
I found my issue with this :)
It currently draws all molecules beforehand, while after the second it can already see that it invalidated the assumption. This snippet would draw a list of unique elements checking the assumption one by one. Lastly, I guess we can't use the `unique` argument for `lists` strategies because that requires the items to be hashable?

```python
def unique_list(draw, strategy, min_size=0, max_size=None):
    output = []
    num_items = draw(st.integer(min=min_size, max=max_size)
    for _ in range(num_items):
        new_item = draw(strategy)
        assume(not new_item  in output)
        output.append(item)
    return output
```

Reply:
I think it is a good idea. Though, I tried and it is slower enough that the health checks start failing.

Reply:
Ok, that's surprising. Leave it as-is then.